File: python/validsudoku.py
from typing import List


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row_set = [set() for _ in range(9)]
        column_set = [set() for _ in range(9)]
        box_set = [set() for _ in range(9)]

        for i in range(9):
            for j in range(9):
                box_index = (i // 3) * 3 + (j // 3)
                if board[i][j] == ".":
                    continue
                elif (
                    board[i][j] in row_set[i]
                    or board[i][j] in column_set[j]
                    or board[i][j] in box_set[box_index]
                ):
                    return False
        return True


sol = Solution()
print(
    sol.isValidSudoku(
        [
            ["1", "2", ".", ".", "3", ".", ".", ".", "."],
            ["4", ".", ".", "5", ".", ".", ".", ".", "."],
            [".", "9", "8", ".", ".", ".", ".", ".", "3"],
            ["5", ".", ".", ".", "6", ".", ".", ".", "4"],
            [".", ".", ".", "8", ".", "3", ".", ".", "5"],
            ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
            [".", ".", ".", ".", ".", ".", "2", ".", "."],
            [".", ".", ".", "4", "1", "9", ".", ".", "8"],
            [".", ".", ".", ".", "8", ".", ".", "7", "9"],
        ]
    )
)

File: c++/validsudoku.cpp
class Solution {
public:
  bool isValidSudoku(vector<vector<char>> &board) {
    vector<unordered_set<char>> rows_arr(9);
    vector<unordered_set<char>> cols_arr(9);
    vector<unordered_set<char>> box_arr(9);
    int rows = board.size();
    int cols = board[0].size();
    // rows
    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < cols; j++) {
        if (board[i][j] == '.') {
          continue;
        }
        if (rows_arr[i].count(board[i][j])) {
          return false;
        }
        rows_arr[i].insert(board[i][j]);
      }
    }
    // cols
    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < cols; j++) {
        if (board[j][i] == '.') {
          continue;
        }
        if (cols_arr[i].count(board[j][i])) {
          return false;
        }
        cols_arr[i].insert(board[j][i]);
      }
    }
    // box

    //(0,0), (0,1), (0,2), (1,0), (1,1),(1,2), (2,0),(2,1),(2,2)
    int box;
    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < cols; j++) {
        box = (j / 3) + (i / 3) * 3;
        if (board[i][j] == '.') {
          continue;
        }
        if (box_arr[box].count(board[i][j])) {
          return false;
        }
        box_arr[box].insert(board[i][j]);
      }
    }
    return true;
  }
};
